Migrate to Github Actions

mjarvis opened the pull request:
Resolves #484, #486 

Migrates CI to GitHub actions.

TODO:
- [x] Update `.scripts/update-gh-pages` to work with github actions

Commit:
First attempt at github actions workflow

Commit:
Fix scheme name

Commit:
Remove name from codecov run

Commit:
Revert to old codecov uploader

Commit:
Add matrix to run swiftpm test & pod lint

Commit:
update if checks

Commit:
Fix missing closing quote

Commit:
Use single quotes for if checks

Commit:
split jobs

Commit:
Add second matrix, Try new codecov uploader again

Commit:
Fix build and run

Commit:
Update iOS destination

Commit:
Exclude gh-pages branch

Commit:
Move update-docs to a separate workflow

Commit:
Add matrices

Commit:
fix update-docs syntax

Commit:
Remove branches-ignore

this will already only be triggered on the correct branches.

Commit:
Don't use env for matrix variables

Commit:
Pass matrix in string

Commit:
Remove TravisCI configuration

DivineDominion reviewed (approved):
This looks pretty good and the actions do run, so nice work! 

Just looking at the result makes it seem so easy. But the commit history reveals you had to battle with the system, too. I feel you, thinking back to the ReSwift-Thunk GH Action implementation I somehow forced to do something useful :)

(I guess caching would be nice to both speed up the process and of course preserve energy to help fight climate change etc. ;) but we don't run CI that much, so I think this is pretty good.)

Your initial PR comment has an unticked box w.r.t. documentation generation. Doesn't the "Update Docs" workflow now already do this? Or does it just call the script but the script isn't doing it properly?

mjarvis commented:
@DivineDominion The update docs script appears to be heavily dependent on travis-ci environment variables. Haven't looked closely but I'm assuming it will need modifications. Will take a look today.

Commit:
Testing update-docs workflow

Temporarily pushing to `gh-pages-test` to validate without affecting actual docs.

Commit:
Temporarily disable ci runs to test only update-docs workflow

Commit:
Switch update-docs to current branch for testing

Commit:
Remove Jazzy install

Jazzy is pre-installed on the macOS environment as per https://github.com/actions/virtual-environments/blob/main/images/macos/macos-11-Readme.md

Commit:
Remove scheme check, since we're in a separate workflow

Commit:
Fail update-gh-pages if anything fails

Commit:
Use github token in push

Commit:
Adjust pushing configuration

Commit:
add origin

Commit:
Try passing token in

Commit:
Switch to live environment

- Re-enable test workflow
- Switch update-docs to `master` and `gh-pages`.

DivineDominion commented on .scripts/update-gh-pages:
```diff
@@ -71,12 +56,12 @@ pushd "$CHECKOUT_PATH"
+    
     echo " ==> Push changes -> '$REPO'"
-    git push -q "https://$GITHUB_PAGES_TOKEN@github.com/$ORGANISATION/$NAME.git" gh-pages
+    git push -q "https://github-actions:$GITHUB_TOKEN@github.com/$ORGANISATION/$NAME.git" gh-pages
```
That's a good find! 👍

DivineDominion commented on .scripts/update-gh-pages:
```diff
@@ -36,24 +27,18 @@ else
 
-if [ "$TRAVIS_BRANCH" != "" ]; then
-    echo " ==> Install: jazzy"
-    gem install --no-rdoc --no-ri jazzy
```
Sorry if that's a dumb question: How's the binary dependency resolved now if we skip the installation step?

mjarvis replied:
Not dumb. Jazzy is pre-installed as per https://github.com/actions/virtual-environments/blob/main/images/macos/macos-11-Readme.md

DivineDominion reviewed (approved):
Looks good in principle, but I'd still be curious to see if this works :)
